test/common_device_type.py

```diff
@@ -83,18 +105,39 @@
 class DeviceTypeTestBase(TestCase):
     device_type = "generic_device_type"
 
+    # Returns the dtypes the test has requested.
+    # Prefers device-specific dtype specifications over generic ones.
+    @classmethod
+    def _get_dtypes(cls, test):
+        if not hasattr(test, 'dtypes'):
+            return None
+        return test.dtypes.get(cls.device_type, test.dtypes.get('all', None))
+
     # Creates device-specific tests.
     @classmethod
     def instantiate_test(cls, test):
         test_name = test.__name__ + "_" + cls.device_type
 
-        assert not hasattr(cls, test_name), "Redefinition of test {0}".format(test_name)
+        dtypes = cls._get_dtypes(test)
+        if dtypes is None:  # Test has no dtype variants
+            assert not hasattr(cls, test_name), "Redefinition of test {0}".format(test_name)
+
+            @wraps(test)
+            def instantiated_test(self, test=test):
+                return test(self, cls.device_type)
 
-        @wraps(test)
-        def instantiated_test(self, test=test):
-            return test(self, cls.device_type)
+            setattr(cls, test_name, instantiated_test)
+        else:  # Test has dtype variants
+            for dtype in dtypes:
+                dtype_str = str(dtype).split('.')[1]
```

Comment:
nit: probably worth throwing an assert that the dtypes passed in are actually dtypes and that the names are `torch.X`?

Reply:
This is checked in the dtypes decorator:

`assert all(arg in ALL_TYPES for arg in args), "Unknown dtype in {0}".format(str(args))`

The name passed in doesn't matter. Taking the string of any valid variant will canonicalize the naming.

Reply:
sure, but now we have to have a list of all the dtypes which is yet another place we have to update to add a dtype (which isn't useful for tests, as I argued). And there are implicit assumptions about what can actually go in the list (i.e. has to be torch.X, otherwise you get either weird or duplicate names).  Instead, we can just trivially and explicitly check the constraints we actually have.